pocs/linux/kernelctf/CVE-2025-21700_lts_cos_mitigation/docs/exploit.md

```diff
@@ -0,0 +1,408 @@
+
+Because of `CONFIG_RANDOM_KMALLOC_CACHES` we cannot easily reclaim the `struct Qdisc`
+object as we did in the LTS versions.
+For now, we will instead focus on the `struct drr_class` object. It turns out
+that we can indeed deterministically reclaim one of the class objects with objects
+of different types because of a similar issue that I pointed out [previously](https://github.com/google/security-research/blob/master/pocs/linux/kernelctf/CVE-2023-6817_mitigation/docs/exploit.md#mitigation-notes):
+Consider the following assembly generated for the class allocation in
+`drr_change_class()`:
+```objdump
+ffffffff81ead050 <drr_change_class>:
+ffffffff81ead050:       e8 8b a0 26 ff          call   ffffffff811170e0 <__fentry__>
+ffffffff81ead055:       41 57                   push   %r15
+ffffffff81ead057:       41 56                   push   %r14
+ffffffff81ead059:       41 55                   push   %r13
+ffffffff81ead05b:       41 54                   push   %r12
+ffffffff81ead05d:       4d 89 cc                mov    %r9,%r12
+ffffffff81ead060:       55                      push   %rbp
+ffffffff81ead061:       53                      push   %rbx
+ffffffff81ead062:       48 83 ec 28             sub    $0x28,%rsp
+ffffffff81ead066:       48 8b 51 10             mov    0x10(%rcx),%rdx
+ffffffff81ead06a:       4d 8b 30                mov    (%r8),%r14
+ffffffff81ead06d:       89 74 24 04             mov    %esi,0x4(%rsp)
+ffffffff81ead071:       65 48 8b 04 25 28 00    mov    %gs:0x28,%rax
+...
+ffffffff81ead1a6:       48 8b 44 24 58          mov    0x58(%rsp),%rax          # [5]
+ffffffff81ead1ab:       48 33 05 c6 ef 32 01    xor    0x132efc6(%rip),%rax     # ffffffff831dc178 <random_kmalloc_seed>
+ffffffff81ead1b2:       be c0 0d 00 00          mov    $0xdc0,%esi
+ffffffff81ead1b7:       48 ba eb 83 b5 80 46    movabs $0x61c8864680b583eb,%rdx
+ffffffff81ead1be:       86 c8 61
+ffffffff81ead1c1:       48 0f af c2             imul   %rdx,%rax
+ffffffff81ead1c5:       48 c1 e8 3c             shr    $0x3c,%rax
+ffffffff81ead1c9:       48 8d 14 c5 00 00 00    lea    0x0(,%rax,8),%rdx
+ffffffff81ead1d0:       00
+ffffffff81ead1d1:       48 29 c2                sub    %rax,%rdx
+ffffffff81ead1d4:       48 89 d0                mov    %rdx,%rax
+ffffffff81ead1d7:       ba 70 00 00 00          mov    $0x70,%edx
+ffffffff81ead1dc:       48 c1 e0 04             shl    $0x4,%rax
+ffffffff81ead1e0:       48 8b b8 b8 c1 1d 83    mov    -0x7ce23e48(%rax),%rdi
+ffffffff81ead1e7:       e8 e4 39 4d ff          call   ffffffff81380bd0 <kmalloc_trace>
+...
+```
+We can see, that the kmalloc call was _inlined_. Additionally, note that the
+return address to compute the random seed is directly fetched from the setup
+function frame ([5]). This is problematic because this function is being called
+by a function pointer, i.e. indirect call, via `struct Qdisc_ops`. This means we
+have access to all `ops->change_class` functions for potential victim allocations
```

Comment:
"have access to all `ops->change_class` functions" -> you mean `sch->ops->change` within `qdisc_change`, right?

Can you be more specific in the writeup?

Reply:
Corrected it, it's the `cl_ops` variant, so not in `qdisc_change`